lib/ansible/modules/identity/keycloak/keycloak_scope.py

```diff
@@ -0,0 +1,472 @@
+            for role in clientroles[client]:
+                checkresult, msg = check_role_representation(role)
+                if not checkresult:
+                    module.fail_json(msg=msg)
+
+    # obtain ID of client or client template when given client_id or name
+    cid = module.params.get('id')
+    if cid is None:
+        if target == 'client':
+            if module.params.get('client_id') is not None:
+                cid = kc.get_client_id(module.params.get('client_id'), realm=realm)
+            else:
+                module.fail_json(msg='When target is client, you need to specify an id or client_id.')
+        else:
+            if module.params.get('name') is not None:
+                cid = kc.get_client_template_id(module.params.get('name'), realm=realm)
+            else:
+                module.fail_json(msg='When target is client-template, you need to specify an id or a name.')
+    if cid is None:
+        module.fail_json(msg='Could not obtain valid client(-template) id to work on.')
+
+    if roletype == 'realm':
+        # in this case, the loop below needs to be run once with a None item
+        scope_clients = [None]
+    if roletype == 'client':
+        scope_clients = list(clientroles.keys())
+        if state == 'exclusive':
+            # when exclusive is set, we want to delete already existing scope mappings regarding all
+            # clients, so add existing mappings' client-names to list of clients to iterate over
+            # (with an empty scope mapping)
+            currentmappings = kc.get_scope_mappings(cid, target=target, realm=realm)
+            if 'clientMappings' in currentmappings:
+                currentmappings = currentmappings['clientMappings']
+            else:
+                currentmappings = dict()
+
+            for mapping in currentmappings.keys():
+                if mapping not in scope_clients:
+                    scope_clients.append(mapping)
+
+    # Let's do some bookkeeping
+    removed_roles = 0
+    updated_roles = 0
+    for scope_client in scope_clients:
+        if scope_client is not None:
+            # If this is a client scope mapping as opposed to a realm scope mapping
+            if scope_client in clientroles:
+                roles = clientroles[scope_client]
+            else:
+                roles = []
+            scope_client_id = kc.get_client_id(scope_client, realm=realm)
+        else:
+            scope_client_id = scope_client
+
+        if roles is None:
+            roles = []
+
+        # obtain current roles
+        current = kc.get_scope_mapping(cid, id_client=scope_client_id, target=target, realm=realm)
+        proposed = deepcopy(current)
+        end_state = []
+
+        to_delete = []
+        if state == 'absent':
+            for role in current:
+                if contains_role(roles, role['name']):
+                    to_delete.append(role)
+                    proposed = [x for x in proposed if x['name'] != role['name']]
```

Comment:
proposed gets overwritten with itself less the role being removed. It could end up being [] if all roles in the mapping are removed, but there may be roles in the mapping server-side that are not mentioned in the module invocation, so it may not be empty at the end.